File: AspNetCore/MCRCOAutomoviles/AspNetCore/Models/Rentals/McrcoAutomovilesModel.cs
//McrcoAutomovilesModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.OData.ModelBuilder;
using Microsoft.EntityFrameworkCore;

namespace MilesCarRental.Rentals.Models.v1
{
    [Table("Automoviles", Schema = "MCRCO")]
    public class McrcoAutomoviles : IDisposable
	{
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int McrcoAutomovilesId {get; set;}

        //[Column(TypeName = "Date")]
        public DateTime McrcoAutomovilesFechaIngreso {get; set;}

        [MaxLength(50)]
        public string McrcoAutomovilesMarca {get; set;}

        [MaxLength(8)]
        public string McrcoAutomovilesModelo {get; set;}

        public int McrcoAutomovilesAno {get; set;}

        [MaxLength(20)]
        public string McrcoAutomovilesClase {get; set;}

        public int McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar {get; set;}

        public int McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega {get; set;}

        public Enum_MCRCOAutomovilesEstado McrcoAutomovilesEstado {get; set;}

        public McrcoSucursales McrcoSucursalesReclamar { get; set; }

        public McrcoSucursales McrcoSucursalesEntrega { get; set; }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        private bool disposed = false;
        
        protected virtual void Dispose(bool disposing)
        {
            if(!this.disposed)
            {
                disposed = true;
            }
        }
	}

    #region Enums
    public enum Enum_MCRCOAutomovilesEstado : int
    {
        Disponible = 1,
        Mantenimiento = 2,
        Fuera_de_servicio = 3
    }
    #endregion


    /// <summary>
    /// Extensión para registrar mapping con el Entity Framework y oData
    /// </summary>
	public static class McrcoAutomovilesExtension
	{
        #region EF Mapping
        public static ModelBuilder McrcoAutomovilesMapping(this ModelBuilder modelBuilder)
        {
            var entity = modelBuilder.Entity<McrcoAutomoviles>();

            //PrimaryKey
            entity.HasKey(c => new { c.McrcoAutomovilesId });

           //Relationships
            entity.HasOne(typeof(McrcoSucursales), "McrcoSucursalesReclamar")
                .WithMany()
                .HasForeignKey("McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar")
                .OnDelete(DeleteBehavior.Restrict); // no ON DELETE

            entity.HasOne(typeof(McrcoSucursales), "McrcoSucursalesEntrega")
                .WithMany()
                .HasForeignKey("McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega")
                .OnDelete(DeleteBehavior.Restrict); // no ON DELETE


			return modelBuilder;
		}

        /// <summary>
        /// Mapping con el oData Framework
        /// </summary>
        /// <param name="oDataModelBuilder"></param>
        public static void McrcoAutomovilesMapping(this ODataConventionModelBuilder oDataModelBuilder)
        {
            var entityConfig = oDataModelBuilder.EntitySet<McrcoAutomoviles>(nameof(McrcoAutomoviles));

            var entity = entityConfig.EntityType;

            // PrimaryKey
            entity.HasKey(c => new { c.McrcoAutomovilesId });



            //Ignored properties for oData
            // entityConfig.EntityType.Ignore(x => x.Summary);
        }
        #endregion
    }
}

File: Angular/automoviles-mcr/src/app/rentals/mcrco.automoviles/mcrco.automoviles.model.ts
export enum EnumMcrcoAutomovilesEstado {
  'Disponible' = '1',
  'Mantenimiento' = '2',
  'Fuera_de_servicio' = '3'
}

export class McrcoAutomovilesModel {
    public McrcoAutomovilesId: number = 0;
    public McrcoAutomovilesFechaIngreso: Date = new Date();
    public McrcoAutomovilesMarca: string;
    public McrcoAutomovilesModelo: string;
    public McrcoAutomovilesAno: number;
    public McrcoAutomovilesClase: string;
    public McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar: number;
    public McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega: number;
    public McrcoAutomovilesEstado: string = 'Disponible';
    public McrcoSucursalesReclamar?: any = {};
    public McrcoSucursalesEntrega?: any = {};
    public _secuencia?: number = 0;
    public _estado?: string = 'N';
    public _id?: string = '';
    public _v?: number = 0;

    constructor(json?: any) {
        if(json) {
            this.McrcoAutomovilesId = json.McrcoAutomovilesId;
            this.McrcoAutomovilesFechaIngreso = json.McrcoAutomovilesFechaIngreso;
            this.McrcoAutomovilesMarca = json.McrcoAutomovilesMarca;
            this.McrcoAutomovilesModelo = json.McrcoAutomovilesModelo;
            this.McrcoAutomovilesAno = json.McrcoAutomovilesAno;
            this.McrcoAutomovilesClase = json.McrcoAutomovilesClase;
            this.McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar = json.McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar;
            this.McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega = json.McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega;
            this.McrcoAutomovilesEstado = json.McrcoAutomovilesEstado;
            this.McrcoSucursalesReclamar = json.McrcoSucursalesReclamar;
            this.McrcoSucursalesEntrega = json.McrcoSucursalesEntrega;
        }
    }

    static clone(row: McrcoAutomovilesModel): McrcoAutomovilesModel {
        const rowCloned = Object.assign({}, row);

        delete rowCloned._secuencia;
        delete rowCloned._estado;
        delete rowCloned._id;
        delete rowCloned._v;
        delete rowCloned.McrcoSucursalesReclamar;
        delete rowCloned.McrcoSucursalesEntrega;

        return rowCloned;
    }

    static cloneExcel(data: McrcoAutomovilesModel[]): any[] {
       const dataExcel: any[] = [];
       data.forEach(dato => {
           const registro = {

              McrcoAutomovilesId: dato.McrcoAutomovilesId,
              McrcoAutomovilesFechaIngreso: dato.McrcoAutomovilesFechaIngreso,
              McrcoAutomovilesMarca: dato.McrcoAutomovilesMarca,
              McrcoAutomovilesModelo: dato.McrcoAutomovilesModelo,
              McrcoAutomovilesAno: dato.McrcoAutomovilesAno,
              McrcoAutomovilesClase: dato.McrcoAutomovilesClase,
              McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar: dato.McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar,
              McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega: dato.McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega,
              McrcoAutomovilesEstado: dato.McrcoAutomovilesEstado,
              McrcoSucursalesReclamar:  dato.McrcoSucursalesReclamar,
              McrcoSucursalesEntrega:  dato.McrcoSucursalesEntrega

            };
            dataExcel.push(registro);
       });
       return dataExcel;
    }

    toClipboard(separator: string = '\t'): string {
        let result = '';

        result += `${separator}${this.McrcoAutomovilesId}`;
        result += `${separator}${this.McrcoAutomovilesFechaIngreso}`;
        result += `${separator}${this.McrcoAutomovilesMarca}`;
        result += `${separator}${this.McrcoAutomovilesModelo}`;
        result += `${separator}${this.McrcoAutomovilesAno}`;
        result += `${separator}${this.McrcoAutomovilesClase}`;
        result += `${separator}${this.McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar}`;
        result += `${separator}${this.McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega}`;
        result += `${separator}${this.McrcoAutomovilesEstado}`;

        return result.substring(separator.length);
    }

    fromClipboard(value: string, separator: string = '\t'): McrcoAutomovilesModel {
        const result = value.split(separator);

        this.McrcoAutomovilesId = parseInt(result[0], 10);
        this.McrcoAutomovilesFechaIngreso = new Date(result[1]);
        this.McrcoAutomovilesMarca = result[2];
        this.McrcoAutomovilesModelo = result[3];
        this.McrcoAutomovilesAno = parseInt(result[4], 10);
        this.McrcoAutomovilesClase = result[5];
        this.McrcoSucursalesIdMcrcoSucursalesDescripcionReclamar = parseInt(result[6], 10);
        this.McrcoSucursalesIdMcrcoSucursalesDescripcionEntrega = parseInt(result[7], 10);
        this.McrcoAutomovilesEstado = result[8];

        return this;
    }

}
